### PR TITLE
PDEV-5066: adding batch_merge_array option to allow for bulk inserting elements

### PR DESCRIPTION
**Pull request by tomzagorski**
Completes [PDEV-5066](https://dewsly.atlassian.net/browse/PDEV-5066)

Background
================
This PR adds the ability to bulk insert elements into a sidekiq-grouping queue. This is particularly useful when handling webhook events that may already be batched. For example, SendGrid status webhooks.

Current APM trace:
image

Testing
================
Added specs and tested in staging with a Smart Alerts notification to ~1300 emails. The emails in a "sending" status look to be possibly invalid domains that will eventually bounce.

Screen Shot 2022-10-27 at 4 16 15 PM
Screen Shot 2022-10-27 at 4 16 27 PM
Screen Shot 2022-10-27 at 4 16 35 PM
Screen Shot 2022-10-27 at 4 16 44 PM
Screen Shot 2022-10-27 at 4 17 52 PM
Screen Shot 2022-10-27 at 4 18 06 PM

Risks
================
None that I'm aware of

Author Checklist
-----------------------

- [x] **A1** I added a Jira issue number
- [x] **A2** I added a Background section
- [x] **A3** I added automated tests, or they are not needed (explain why)
- [x] **A4** I described the manual tests I performed for this change
- [x] **A5** I performance tested this with real data, or this does not affect performance
- [x] **A6** I described any risks with this change

Reviewer Checklist
-----------------------

- [ ] **R1** I understand the scope and function of this change
- [ ] **R2** This change fulfills the requirements in the Jira issue
- [ ] **R3** This change is good quality and provides a good user experience
- [ ] **R4** Future maintainers will be able to understand this code
- [ ] **R5** This code is sufficiently tested
- [ ] **R6** This change requires no further verification or testing

### CODE REVIEW

File: lib/sidekiq/grouping/middleware.rb [hunk 1]
```diff
@@ -32,10 +32,11 @@ def call(worker_class, msg, queue, redis_pool = nil)
       private
 
       def add_to_batch(worker_class, queue, msg, redis_pool = nil)
+        add_method = worker_class.get_sidekiq_options['batch_merge_array'] ? :merge : :add
+
         Sidekiq::Grouping::Batch
           .new(worker_class.name, queue, redis_pool)
-          .add(msg['args'])
-
+          .send(add_method, msg['args'])
```
Comment (justinhoward) [inline lib/sidekiq/grouping/middleware.rb]: We can use `public_send` to make it clear we're calling public APIs here.

File: lib/sidekiq/grouping/redis.rb [hunk 2]
```diff
@@ -91,6 +122,12 @@ def push_msg(name, msg, remember_unique = false)
         end
       end
 
+      def push_messages(name, messages, remember_unique = false)
+        keys = [ns('batches'), name, ns(name), unique_messages_key(name), remember_unique]
+        args = [messages]
+        redis { |conn| conn.evalsha @script_hashes[:merge_array], keys, args }
```
Comment (justinhoward) [inline lib/sidekiq/grouping/redis.rb:128]: Running a script is an exclusive operation, and messages is unbounded here. I suggest we set a hard limit on the size of messages and batch them accordingly. Sidekiq does a similar thing with `push_bulk` (although it doesn't use lua scripting for pushes). If we don't batch, an unbounded push will block all redis requests for an indeterminate amount of time.

https://github.com/mperham/sidekiq/blob/main/lib/sidekiq/job.rb#L240-L247

File: lib/sidekiq/grouping/redis.rb [hunk 3]
```diff
@@ -61,19 +61,50 @@ class Redis
         redis.call('zrem', pending_jobs, expired_queue)
       LUA
 
+      MERGE_ARRAY_SCRIPT = <<-LUA
```
Comment (justinhoward) [inline lib/sidekiq/grouping/redis.rb:64]: This looks great. Definitely a nice way to optimize.

File: lib/sidekiq/grouping/redis.rb (hunk 2 above)
Comment (tomzagorski) [reply lib/sidekiq/grouping/redis.rb:128]: Great point. Thanks for pointing this out. I put the batching of these inserts up a level at the batch itself but that wouldn't stop someone from misusing it here. Going to push another commit to move that here instead soon.

**Review by justinhoward — APPROVED** [review]
Nice!

**Pull request merged**